jupyterdrive/gdrive/drive-contents.js

```diff
@@ -87,6 +52,136 @@ define(function(require) {
+
+    ClobberDetector.prototype.add_uuid_to_cache = function(resource, uuid) {
+        var file_id = resource['id'];
+        this.save_uuids[file_id] = this.save_uuids[file_id] || [];
+        var uuids = this.save_uuids[file_id];
+        uuids.push(uuid);
+        if (uuids.length > ClobberDetector.NUM_UUIDS_IN_CACHE) {
+            uuids.shift();
+        }
+        console.log(uuids);
+    }
+
+    /**
+     * Returns a UUID
+     * @return {string}
+     */
+    ClobberDetector.prototype.create_uuid = function() {
+        return Math.random().toString().substr(2);
+    }
+
+    /**
+     * Returns file metadata that sets a custom property to record
+     * the provided uuid as the id of the last save
+     * @return {Object}
+     */
+    ClobberDetector.prototype.create_metadata = function(uuid) {
+        return {
+            'properties' : [{
+                'visibility': 'PRIVATE',  // only visible to this app
+                'key': ClobberDetector.SAVE_UUID_KEY,
+                'value': uuid
+            }]
+        };
+    };
+
+    /**
+     * Should be called whenever a file loaded.  Since the file being
+     * loaded is the one that will populate this document, the SAVE_ID
+     * of this file will be assumed not to indicate cloberring.
+     * @param {Object} resource The Fiels resource of the file that was
+     *     loaded.
+     */
+    ClobberDetector.prototype.on_load_file = function(resource) {
+        // Go through all properties, searching for the relevant property,
+        // and record it if it exists.
+        var that = this;
+        if (!resource['properties']) { return; }
+        resource['properties'].forEach(function(property) {
+            if (property['visibility'] === 'PRIVATE'
+                && property['key']=== ClobberDetector.SAVE_UUID_KEY) {
+                that.add_uuid_to_cache(resource, property['value']);
+            }
+        });
+    };
+
+    /**
+     * Should be called whenever a new version of the file is uploaded
+     * to drive.  Returns metadata that should be uploaded along with
+     * the file.  This metadata identifies the last save.
+     * @param {Object} resource The Files resource of the file being saved
+     * @return {Object} metdata to upload with the file.
+     */
+    ClobberDetector.prototype.on_upload_new_version = function(resource) {
+        var uuid = this.create_uuid();
+        this.add_uuid_to_cache(resource, uuid);
+        return this.create_metadata(uuid);
+    };
+
+    /**
+     * Detects whether a file has been clobbered by another user
+     * @return {Bool} True if clobbering is detected.
+     */
+    ClobberDetector.prototype.has_clobber_occurred = function(resource) {
+        console.log(this.save_uuids);
+        var uuids = this.save_uuids[resource['id']];
+        // In the unlikely event that there are no uuid's for this file
+        // assume no clobbering has occurred.
+        if (!uuids) { return false; }
```

Comment:
wouldn't that be the opposite ? If something else saves, then there has been clobbering ?